File: linda-presentation/src/main/java/it/unibo/coordination/linda/presentation/Deserializer.java
package it.unibo.coordination.linda.presentation;

import java.io.Reader;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static java.util.Collections.singletonList;

public interface Deserializer<T> {

    Class<T> getSupportedType();

    MIMETypes getSupportedMIMEType();

    T fromDynamicObject(Object dynamicObject);

    default List<T> listFromDynamicObject(Object dynamicObject) {
        if (!(dynamicObject instanceof List) && !(dynamicObject instanceof Map)) {
            throw new IllegalArgumentException();
        }

        List<?> list = (dynamicObject instanceof List) ? (List<?>) dynamicObject : singletonList((Map<?, ?>) dynamicObject);

        return list.stream().map(this::fromDynamicObject).collect(Collectors.toList());
    }

    T fromString(String string);

    List<T>  listFromString(String string);

    T read(Reader reader);

    List<T>  readList(Reader reader);

    static <X> Deserializer<X> of(Class<X> klass, MIMETypes mimeType) {
        return Presentation.getDeserializer(klass, mimeType);
    }
}

File: presentation/src/main/kotlin/it/unibo/presentation/Deserializer.kt
package it.unibo.presentation

import java.io.Reader

interface Deserializer<T> {
    val supportedType: TypeToken<T>

    val supportedMIMEType: MIMETypes

    fun fromDynamicObject(dynamicObject: Any): T

    @JvmDefault
    fun listFromDynamicObject(dynamicObject: Any): List<T> {
        return when(dynamicObject) {
            is Map<*, *> -> listOf(fromDynamicObject(dynamicObject))
            is List<*> -> dynamicObject.map { fromDynamicObject(it!!) }
            else -> throw IllegalArgumentException()
        }
    }

    fun fromString(string: String): T

    fun listFromString(string: String): List<T>

    fun read(reader: Reader): T

    fun readList(reader: Reader): List<T>

}